rewriter/flatten.cc

```diff
@@ -166,10 +168,34 @@ class FlattenWalk {
     // need to move to the end, so we keep that below on the stack.
     vector<ClassScope> classScopes;
 
-    int computeStaticLevel(bool isSelf) {
+    // compute the new scope information.
+    ScopeInfo computeScopeInfo(ScopeType scopeType) {
         auto &methods = curMethodSet();
-        int prevLevel = methods.stack.empty() ? 0 : methods.stack.back().staticLevel;
-        return prevLevel + (isSelf ? 1 : 0);
+        // if we're in instance scope, then we'll always stay in it. This is an incorrect approximation, but it's as
+        // close as we can get with our specific model: in particular, if we have something like
+        //
+        // def foo; def bar; end; end
+        //
+        // then both #foo and #bar will be instance methods, but if we have
+        //
+        // def foo; def self.bar; end; end
+        //
+        // then #foo will be an instance method while #bar will be an instance method stored on the singleton class for
```

Comment:
> stored on the singleton class

is it actually stored on the singleton class? i would have thought it was stored on that object individually

are you saying that there is only one definition of `def self.bar`, no matter on how many objects `x.foo` is run?

Reply:
Also, is this different from how `def self.foo; def bar; end; end` and `def self.foo; def self.bar; end; end` work?

Reply:
In Ruby, both classes _and_ instances can have singleton classes, and `bar` in the example there is stored in the singleton class _of the instance_. A convention I've seen (in the [blog post here](http://leohetsch.com/demystifying-ruby-singleton-classes/)) is to use `#Foo` to refer to the singleton class attached to a class and `#foo` to refer to the singleton class attached to the instance.

```ruby
class SomeClass
  def f; end  # method defined on SomeClass
  def self.f; end  # method defined on #SomeClass
end
some_instance = SomeClass.new
def some_instance.f; end # method defined in #some_instance, which
  # is the singleton class associated with the instance
```

When you have a `def` inside of a static method, then it acts basically the same as when you define it outside. e.g. `def self.f; def self.g; end; end` you'll get both methods defined on the singleton class of the context they're in: if they're in a class, then they'll both be defined on the singleton class for that class. In an instance context, though, adding `self` is going to start putting things on the instance's singleton class, and lacking it will simply define another instance method. So, assuming all of these come in the context of `SomeClass`:

```ruby
def self.outer  # defined on #SomeClass 
  def inner; end  # defined on SomeClass
end

def self.outer  # defined on #SomeClass 
  def inner; end  # defined on #SomeClass 
end

def outer  # defined on SomeClass
  def inner; end  # define on SomeClass
end

def outer  # defined on SomeClass
  def self.inner; end  # defined on #some_instance
end
```